types/module/server/router.go

```diff
@@ -0,0 +1,194 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	gogogrpc "github.com/gogo/protobuf/grpc"
+	"google.golang.org/grpc"
+	"reflect"
+)
+
+type handler struct {
+	f            func(ctx context.Context, args, reply interface{}) error
+	commitWrites bool
+	moduleName   string
+}
+
+type router struct {
+	handlers         map[string]handler
+	providedServices map[reflect.Type]bool
+	antiReentryMap   map[string]bool
+	authzMiddleware  AuthorizationMiddleware
+}
+
+type registrar struct {
+	*router
+	baseServer   gogogrpc.Server
+	commitWrites bool
+	moduleName   string
+}
+
+var _ gogogrpc.Server = registrar{}
+
+func (r registrar) RegisterService(sd *grpc.ServiceDesc, ss interface{}) {
+	r.providedServices[reflect.TypeOf(sd.HandlerType)] = true
+
+	r.baseServer.RegisterService(sd, ss)
+
+	for _, method := range sd.Methods {
+		fqName := fmt.Sprintf("/%s/%s", sd.ServiceName, method.MethodName)
+		methodHandler := method.Handler
+		f := func(ctx context.Context, args, reply interface{}) error {
+			res, err := methodHandler(ss, ctx, func(i interface{}) error { return nil },
+				func(ctx context.Context, _ interface{}, _ *grpc.UnaryServerInfo, unaryHandler grpc.UnaryHandler) (resp interface{}, err error) {
+					return unaryHandler(ctx, args)
+				})
+			if err != nil {
+				return err
+			}
+
+			resValue := reflect.ValueOf(res)
+			if !resValue.IsZero() {
+				reflect.ValueOf(reply).Elem().Set(resValue.Elem())
+			}
+			return nil
+		}
+		r.handlers[fqName] = handler{
+			f:            f,
+			commitWrites: r.commitWrites,
+			moduleName:   r.moduleName,
+		}
+	}
+}
+
+func (rtr *router) invoker(methodName string, writeCondition func(context.Context, string, sdk.MsgRequest) error) (Invoker, error) {
+	handler, found := rtr.handlers[methodName]
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, fmt.Sprintf("cannot find method named %s", methodName))
+	}
+
+	moduleName := handler.moduleName
+
+	if writeCondition != nil && handler.commitWrites {
+		// msg handler
+		return func(ctx context.Context, request interface{}, response interface{}, opts ...interface{}) error {
+			if _, alreadyOnCallStack := rtr.antiReentryMap[moduleName]; alreadyOnCallStack {
+				return fmt.Errorf("re-entrant module calls not allowed for security reasons! module %s is already on the call stack", moduleName)
+			}
+
+			rtr.antiReentryMap[moduleName] = true
+			defer delete(rtr.antiReentryMap, moduleName)
```

Comment:
This only works when we won't have async calls in `hander` (if handler is able to make a async call for other module , then this protection breaks apart.

As an alternative, we never delete from the map. Instead we create a new one once the app will start processing a new transaction.

Reply:
Yeah async currently was isn't allowed because there is no consensus way to do ordering.